src/shim.c

```diff
@@ -322,188 +505,133 @@ handle_stdin(struct cc_shim *shim)
-	set_big_endian_64 (buf, shim->io_seq_no);
-	set_big_endian_32 (buf + STREAM_HEADER_LENGTH_OFFSET, (uint32_t)len);
-
-	// TODO: handle write in the poll loop to account for write blocking
-	ret = (int)write(shim->proxy_io_fd, buf, (size_t)len);
-	if (ret == -1) {
-		shim_warning("Error writing from fd %d to fd %d: %s\n",
-			     STDIN_FILENO, shim->proxy_io_fd, strerror(errno));
-		return;
-	}
+	send_proxy_message(shim, type_stream, stream_stdin, buf);
 }
 
 /*!
- * Read and parse I/O message on proxy I/O fd
- *
- * \param shim \ref cc_shim
- * \param[out] seq Seqence number of the I/O stream
- * \param[out] stream_len Length of the data received
+ * Handle response received from proxy
  *
- * \return newly allocated string on success, else \c NULL.
+ *\param fr \ref frame
  */
-char*
-read_IO_message(struct cc_shim *shim, uint64_t *seq, ssize_t *stream_len) {
-	char *buf = NULL;
-	ssize_t need_read = STREAM_HEADER_SIZE;
-	ssize_t bytes_read = 0, want, ret;
-	ssize_t max_bytes = HYPERSTART_MAX_RECV_BYTES;
-
-	if (! (shim && seq && stream_len)) {
-		return NULL;
+void
+handle_proxy_response(struct frame *fr)
+{
+	if ( ! (fr && fr->payload)) {
+		return;
 	}
 
-	*stream_len = 0;
-
-	buf = calloc(STREAM_HEADER_SIZE, 1);
-	if (! buf ) {
-		abort();
+	// Reponses received from proxy are currently just logged.
+	if (fr->header.err) {
+		shim_error("Error response received from proxy: %s\n", fr->payload);
+	} else {
+		/* TODO: Currently logging response. Do we want to track responses
+		 * to requests in future. Maybe useful for restarting connection with proxy
+		 * and resending requests that were missed.
+		 */
+		shim_debug("Response received from proxy: %s\n", fr->payload);
 	}
+}
 
-	while (bytes_read < need_read) {
-		want = need_read - bytes_read;
-		if (want > BUFSIZ)  {
-			want = BUFSIZ;
-		}
-
-		ret = read(shim->proxy_io_fd, buf+bytes_read, (size_t)want);
-		if (ret == -1) {
-			free(buf);
-			err_exit("Error reading from proxy I/O fd: %s\n", strerror(errno));
-		} else if (ret == 0) {
-			/* EOF received on proxy I/O fd*/
-			free(buf);
-			err_exit("EOF received on proxy I/O fd\n");
-		}
-
-		bytes_read += ret;
-
-		if (*stream_len == 0 && bytes_read >= STREAM_HEADER_SIZE) {
-			*stream_len = get_big_endian_32((uint8_t*)(buf+STREAM_HEADER_LENGTH_OFFSET));
+/*!
+ * Handle stream received from proxy
+ *
+ *\param fr \ref frame
+ */
+void
+handle_proxy_stream(struct frame *fr)
+{
+	int outfd;
+	ssize_t offset = 0;
+	size_t ret;
 
-			// length is 12 when hyperstart sends eof before sending exit code
-			if (*stream_len == STREAM_HEADER_SIZE) {
-				break;
-			}
+	if ( ! (fr && fr->payload)) {
+		return;
+	}
 
-			/* Ensure amount of data is within expected bounds */
-			if (*stream_len > max_bytes) {
-				shim_warning("message too big (limit is %lu, but proxy returned %lu)",
-						(unsigned long int)max_bytes,
-						(unsigned long int)stream_len);
-				goto err;
-			}
+	if (fr->header.opcode == stream_stdout) {
```

Comment:
Since there are multiple opcodes with the same value, I'd perform a check on `fr->header.type` at the top of this function to assert that you are actually operating on a `type_stream` frame.